src/bin/wc.rs

```diff
@@ -0,0 +1,149 @@
+use std::env;
+use std::fs::File;
+use std::io;
+use std::io::BufReader;
+use std::io::Read;
+
+struct Config {
+    bytes: bool,
+    lines: bool,
+    chars: bool,
+    words: bool,
+}
+
+struct Output {
+    bytes: u32,
+    lines: u32,
+    chars: u32,
+    words: u32,
+    filename: Option<String>,
+}
+
+impl Config {
+    pub fn new() -> Config {
+        let (bytes, lines, chars, words) = (false, false, false, false);
+        Config {
+            bytes,
+            lines,
+            chars,
+            words,
+        }
+    }
+    pub fn default(&mut self) -> () {
+        self.bytes = true;
+        self.lines = true;
+        self.words = true;
+    }
+}
+
+fn process<T: io::Read>(
+    reader: BufReader<T>,
+    filename: Option<String>,
+) -> Result<Output, io::Error> {
+    let (mut bytes, mut lines, chars, mut words) = (0, 0, 0, 0);
+    let mut whitespace = true;
+    for byte in reader.bytes() {
+        match byte {
+            Ok(b'\n') => {
+                whitespace = true;
+                lines = lines + 1;
+            }
+            Ok(b'\t') | Ok(b' ') => {
+                whitespace = true;
+            }
+            Err(x) => return Err(x),
+            _ => {
+                if whitespace {
+                    whitespace = false;
+                    words = words + 1;
+                }
+                ()
+            }
+        };
+        bytes = bytes + 1;
+    }
+    Ok(Output {
+        bytes,
+        lines,
+        chars,
+        words,
+        filename,
+    })
+}
+
+fn display(config: &Config, output: &Output) -> String {
+    let mut res = String::new();
+    if config.lines {
+        res.push_str(&format!("{:>8}", output.lines));
+    }
+    if config.words {
+        res.push_str(&format!("{:>8}", output.words));
+    }
+    if config.bytes {
+        res.push_str(&format!("{:>8}", output.bytes));
+    }
+    if config.chars {
+        res.push_str(&format!("{:>8}", output.chars));
+    }
+
+    match &output.filename {
+        Some(x) => res.push_str(&format!(" {:<}", x)),
+        _ => (),
+    }
+    res
+}
+
+fn main() -> Result<(), io::Error> {
+    let args: Vec<String> = env::args().collect();
```

Comment:
rustness: I'd recommend using an existing commandline parser instead of rolling your own, though I'm sure this is good practice. [Clap](https://github.com/clap-rs/clap) is the big famous one also [mentioned in the rust cookbook](https://rust-lang-nursery.github.io/rust-cookbook/cli/arguments.html), but there's also several struct-based parsers (eg. [structopt](https://docs.rs/structopt/)).

Reply:
Yes, thanks for the tip! For this, I wanted to get more experience with Rust. I will try Clap and other parsers for future exercises!